packages/language/src/language-server/document-symbol-builder.ts

```diff
@@ -0,0 +1,326 @@
+  ): DocumentSymbol[];
+}
+
+class ProcedureSymbolBuilder implements SymbolBuilder {
+  canHandle(token: IToken): boolean {
+    return (
+      token.tokenTypeIdx == CstNodeKind.EntryStatement_Variable &&
+      token.payload?.element.kind === SyntaxKind.ProcedureStatement
+    );
+  }
+
+  buildSymbols(
+    token: IToken,
+    elementTokens: IToken[],
+    textDocument: TextDocument,
+    childSymbols: DocumentSymbol[],
+  ): DocumentSymbol[] {
+    const labelPrefixStatement = token.payload?.element.container as Statement;
+    const procedureName =
+      labelPrefixStatement.labels.map((label) => label.name).join(" ") ??
+      "<unknown>";
+    const range = getSyntaxNodeRange(labelPrefixStatement.labels[0]) ?? {
+      start: 0,
+      end: 0,
+    };
+
+    const symbol = createDocumentSymbol(
+      procedureName,
+      SymbolKind.Function,
+      range,
+      childSymbols,
+      textDocument,
+    );
+
+    symbol.range.end = rangeToLSP(
+      textDocument,
+      tokenToRange(elementTokens[elementTokens.length - 1]),
+    ).end;
+    return [symbol];
+  }
+}
+
+class DeclareSymbolBuilder implements SymbolBuilder {
+  canHandle(token: IToken): boolean {
+    if (
+      token.tokenTypeIdx !== CstNodeKind.DefineOrdinalStatement_Unsigned0 ||
+      token.payload?.element.kind !== SyntaxKind.DeclareStatement
+    ) {
+      return false;
+    }
+
+    // There is still a chance that all variables are generated by the preprocessor.
+    // However, to check this, we would need to traverse the remaining subtree, so we
+    // do the check during symbol creation. No invalid symbols will be created.
+    return true;
+  }
+
+  buildSymbols(
+    token: IToken,
+    elementTokens: IToken[],
+    textDocument: TextDocument,
+    childSymbols: DocumentSymbol[],
+  ): DocumentSymbol[] {
+    const documentSymbols: DocumentSymbol[] = [];
+    const levelSymbols: LevelSymbol[] = [];
+    const hierarchyBuilder = new LevelHierarchyBuilder();
+
+    for (const item of token.payload?.element.items.filter(
+      (i: DeclaredItem) => i.kind === SyntaxKind.DeclaredItem,
+    )) {
+      levelSymbols.push(
+        ...this.retrieveLevelSymbols(item, null, childSymbols, textDocument),
+      );
+    }
+
+    for (const levelSymbol of levelSymbols) {
+      hierarchyBuilder.addSymbol(levelSymbol.level, levelSymbol.symbol);
+    }
+
+    documentSymbols.push(...hierarchyBuilder.getRootSymbols());
+
+    return documentSymbols;
+  }
+
+  private retrieveLevelSymbols(
+    item: DeclaredItem,
+    inheritedLevel: number | null,
+    children: DocumentSymbol[],
+    textDocument: TextDocument,
+  ): LevelSymbol[] {
+    const levelSymbols: LevelSymbol[] = [];
+
+    let kind: SymbolKind = SymbolKind.Variable;
+    if (
+      // Type is lower case in the ast but upper case in the definition.
+      item.attributes.some(
+        (a) =>
+          a.kind === SyntaxKind.ComputationDataAttribute &&
+          a.type?.toString() === "constant",
+      )
```

Comment:
Suggestion: We perform case-insensitive parsing, so this comparison also needs to be done case-insensitive.